### PR TITLE
Promote IOTHubEndpointStorageContainer and IOTHubFallbackRoute to v1alpha2

### PR DESCRIPTION
**Pull request by ytsarev**
### Description of your changes

This PR promotes IOTHubFallbackRoute and its dependency of IOTHubEndpointStorageContainer to v1alpha2

Caveats:
* Special handling of `FallbackRoute/default` externalName/id as it is always the same according to 

#149 should be merged first

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

```sh
k apply -f examples/devices/iothubendpointstoragecontainer.yaml
k get iothubendpointstoragecontainer.devices.azure.jet.crossplane.io/yury-example
NAME           READY   SYNCED   EXTERNAL-NAME   AGE
yury-example   True    True     yury-example    34h
k apply -f examples/devices/iothubfallbackroute.yaml 
NAME           READY   SYNCED   EXTERNAL-NAME   AGE
yury-example   True    True     default         58m
```

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: config/devices/config.go [hunk 1]
```diff
@@ -131,4 +131,42 @@ func Configure(p *config.Provider) {
...
+				Type: "IOTHubEndpointStorageContainer",
+			},
+		}
+		r.UseAsync = true
+		r.ExternalName = config.IdentifierFromProvider
+		r.ExternalName.GetExternalNameFn = common.GetNameFromFullyQualifiedID
+		// /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/mygroup1/providers/Microsoft.Devices/IotHubs/hub1/FallbackRoute/default
+		r.ExternalName.GetIDFn = common.GetFullyQualifiedIDFn("Microsoft.Devices",
+			"IotHubs", "iothub_name",
+			"FallbackRoute", "default")
```
Comment (ulucinar) [inline config/devices/config.go:170]: I would prefer having the exception (regarding the value of `FallbackRoute` being the constant `default`) be implemented here with something like:
```suggestion
		r.ExternalName.GetIDFn = func(ctx context.Context, externalName string, parameters map[string]interface{}, providerConfig map[string]interface{}) (string, error) {
			p, err := common.GetFullyQualifiedIDFn("Microsoft.Devices", "IotHubs", "iothub_name")(ctx, externalName, parameters, providerConfig)
			if err != nil {
				return "", err
			}
			return fmt.Sprintf("%s/FallbackRoute/default", p), nil
		}
```

File: config/common/common.go [hunk 2]
```diff
@@ -129,6 +129,12 @@ func GetFullyQualifiedIDFn(serviceProvider string, keyPairs ...string) tjconfig.
 
 		path := fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/%s", subIDStr, rgStr, serviceProvider)
 		for i := 0; i < len(keyPairs); i += 2 {
+			// https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/iothub_fallback_route#import
+			// FallbackRoute is always default and we don't have associated parameter for it
+			if keyPairs[i] == "FallbackRoute" {
+				path = filepath.Join(path, "FallbackRoute", "default")
+				return path, nil
```
Comment (ulucinar) [inline config/common/common.go:136]: What if in another context `FallbackRoute` is not the last path component or its value is not a constant but instead a value of a parameter? Please see the suggestion below for the configuration of the `azurerm_iothub_fallback_route` resource.

**Review by ulucinar — CHANGES_REQUESTED** [review]

File: config/devices/config.go (hunk 1 above)
Comment (ytsarev) [reply config/devices/config.go:170]: @ulucinar that's what I tried to do first, but GetFullyQualifiedIDFn returns a function, not (string, error). So getting a string out of it is a bit problematic in this context unless I am missing something obvious

Comment (ytsarev) [reply config/devices/config.go:170]: I guess I get it now! Let me test it properly

**Comment by ulucinar** [comment]
> @ulucinar thanks a lot for the suggestion, it worked! Please review again

Hi @ytsarev,
Sure thing! I've merged the parent PR. Could you please rebase this PR?

**Comment by ytsarev** [comment]
@ulucinar rebased, thanks!

File: config/devices/config.go [hunk 3]
```diff
@@ -131,4 +134,48 @@ func Configure(p *config.Provider) {
 		r.ExternalName.GetIDFn = common.GetFullyQualifiedIDFn("Microsoft.Devices", "IotHubs", "iothub_name",
 			"IotHubKeys", "name")
 	})
+
+	p.AddResourceConfigurator("azurerm_iothub_endpoint_storage_container", func(r *config.Resource) {
+		r.Version = common.VersionV1Alpha2
+		r.References = config.References{
+			"iothub_name": config.Reference{
```
Comment (ulucinar) [inline config/devices/config.go:141]: Looks like `iothub_name` argument is [deprecated](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/iothub_endpoint_storage_container#iothub_name) in favor of `iothub_id`. We can add a referencer for the `iothub_id` argument using [`rconfig.ExtractResourceID`](https://github.com/crossplane-contrib/provider-jet-azure/blob/ebac3f7f2f4d8b9e89e255d0d76c7d703df4cd1f/apis/rconfig/config.go#L51) as the extractor.

File: apis/devices/v1alpha2/zz_iothubfallbackroute_types.go [hunk 4]
```diff
@@ -37,11 +37,25 @@ type IOTHubFallbackRouteParameters struct {
...
-	IOTHubName *string `json:"iothubName" tf:"iothub_name,omitempty"`
+	// +kubebuilder:validation:Optional
+	EndpointNamesRefs []v1.Reference `json:"endpointNamesRefs,omitempty" tf:"-"`
+
+	// +kubebuilder:validation:Optional
+	EndpointNamesSelector *v1.Selector `json:"endpointNamesSelector,omitempty" tf:"-"`
+
+	// +crossplane:generate:reference:type=IOTHub
+	// +kubebuilder:validation:Optional
+	IOTHubName *string `json:"iothubName,omitempty" tf:"iothub_name,omitempty"`
```
Comment (ulucinar) [inline apis/devices/v1alpha2/zz_iothubfallbackroute_types.go:52]: It's interesting that we have generated this field's serialized name as `iothubName` but the corresponding ref field's and selector field's serialized names as `iotHubNameRef` and `iotHubNameSelector`, respectively.

File: config/devices/config.go (hunk 3 above)
Comment (ulucinar) [reply config/devices/config.go:141]: Let's not risk the planned delivery with a native provider version bump then. Thank you @ytsarev for checking it.

**Review by ulucinar — APPROVED** [review]
Thank you @ytsarev, very much appreciated :)

**Pull request merged**